jboss/container/eap/prometheus/config/7.1/artifacts/opt/jboss/container/prometheus/etc/jmx-exporter-config.yaml

```diff
@@ -0,0 +1,354 @@
+  labels:
+    deployment: $1
+    subdeployment: $2
+  attrNameSnakeCase: true
+  type: GAUGE
+- pattern: "jboss.as<deployment=(.+), *subsystem=undertow><>(expired_sessions|rejected_sessions|sessions_created): (.+)"
+  name: undertow_sessions_$2
+  help: Undertow Session Metrics $2
+  value: $3
+  valueFactor: 1
+  labels:
+    deployment: $1
+  attrNameSnakeCase: true
+  type: COUNTER
+- pattern: "jboss.as<deployment=(.+), *subdeployment=(.+), *subsystem=undertow><>(expired_sessions|rejected_sessions|sessions_created): (.+)"
+  name: undertow_sessions_$3
+  help: Undertow Session Metrics $3
+  value: $4
+  valueFactor: 1
+  labels:
+    deployment: $1
+    subdeployment: $2
+  attrNameSnakeCase: true
+  type: COUNTER
+# Servlets
+- pattern: "jboss.as<deployment=(.+), *subsystem=undertow, *servlet=(.+)><>(max_request_time|min_request_time): (.+)"
+  name: undertow_servlets_$3_seconds
+  help: Undertow Servlet Metrics $3
+  value: $4
+  valueFactor: .001
+  labels:
+    deployment: $1
+    servlet: $2
+  attrNameSnakeCase: true
+  type: GAUGE
+- pattern: "jboss.as<deployment=(.+), *subdeployment=(.+), *subsystem=undertow, *servlet=(.+)><>(max_request_time|min_request_time): (.+)"
+  name: undertow_servlets_$4_seconds
+  help: Undertow Servlet Metrics $4
+  value: $5
+  valueFactor: .001
+  labels:
+    deployment: $1
+    subdeployment: $2
+    servlet: $3
+  attrNameSnakeCase: true
+  type: GAUGE
+- pattern: "jboss.as<deployment=(.+), *subsystem=undertow, *servlet=(.+)><>(total_request_time): (.+)"
+  name: undertow_servlets_$3_seconds
+  help: Undertow Servlet Metrics $3
+  value: $4
+  valueFactor: .001
+  labels:
+    deployment: $1
+    servlet: $2
+  attrNameSnakeCase: true
+  type: COUNTER
+- pattern: "jboss.as<deployment=(.+), *subdeployment=(.+), *subsystem=undertow, *servlet=(.+)><>(total_request_time): (.+)"
+  name: undertow_servlets_$4_seconds
+  help: Undertow Servlet Metrics $4
+  value: $5
+  valueFactor: .001
+  labels:
+    deployment: $1
+    subdeployment: $2
+    servlet: $3
+  attrNameSnakeCase: true
+  type: COUNTER
+- pattern: "jboss.as<deployment=(.+), *subsystem=undertow, *servlet=(.+)><>(request_count): (.+)"
+  name: undertow_servlets_$3
+  help: Undertow Servlet Metrics $3
+  value: $4
+  valueFactor: 1
+  labels:
+    deployment: $1
+    servlet: $2
+  attrNameSnakeCase: true
+  type: COUNTER
+- pattern: "jboss.as<deployment=(.+), *subdeployment=(.+), *subsystem=undertow, *servlet=(.+)><>(request_count): (.+)"
+  name: undertow_servlets_$4
+  help: Undertow Servlet Metrics $4
+  value: $5
+  valueFactor: 1
+  labels:
+    deployment: $1
+    subdeployment: $2
+    servlet: $3
+  attrNameSnakeCase: true
+  type: COUNTER
+
+# EJB
+- pattern: "jboss.as<deployment=(.+), *subsystem=ejb3, *(.+)=(.+)><>(peak_concurrent_invocations|pool_available_count|pool_current_size|pool_max_size): (.+)"
+  name: ejbs_$4
+  help: EJB Metrics $4
+  value: $5
+  valueFactor: 1
+  labels:
+    deployment: $1
+    type: $2
+    bean: $3
```

Comment:
I agree that it would be better to be consistent with other patterns and have different labels for the 4 children of http://wildscribe.github.io/WildFly/14.0/deployment/subsystem/ejb3/index.html

I am currently adding to the metrics exposed by the WildFly subsystem in https://github.com/wildfly/wildfly/pull/11847 so that it maps to the labels you propose. Once this PR is merged, the metric [message-driven-bean's peak-concurrent-invocations](
http://wildscribe.github.io/WildFly/14.0/deployment/subsystem/ejb3/message-driven-bean/index.html#attr-peak-concurrent-invocations) will be tagged with:

```
deployment=example.war
subsystem=ejb3
message-driven-bean= X
attribute=peak-concurrent-invocations
```

The name of the Prometheus metric is a mouthful though: `deployment_example_war_subsystem_ejb3_message_driven_bean_x_peak_concurrent_invocations`

Reply:
I can make these changes, and `$2: $3` is a legal substitution.

That said, @jmesnil, I think it's not very prometheusy to include labely things in the metric name.  I think the idea here is that the metric name would be `ejb3_peak_concurrent_invocations` with the distinction for deployment, bean type, and bean name being specified on the labels.  I think the idea here is that this allows users to more easily aggregate metrics, in this case `peak_concurrent_invocations` for ejbs (`ejb3_`), assuming I correctly understand the intent.